Library/Homebrew/livecheck/strategy/cpan.rb

```diff
@@ -0,0 +1,67 @@
+# typed: false
+# frozen_string_literal: true
+
+require "uri"
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      # The {Cpan} strategy identifies versions of software at
+      # cpan.metacpan.org by checking directory listing pages.
+      #
+      # CPAN URLs take the following format:
+      #
+      # * `https://cpan.metacpan.org/authors/id/M/MI/MIYAGAWA/Carton-v1.0.34.tar.gz`
+      #
+      # @api public
+      class Cpan
+        NICE_NAME = "CPAN"
+
+        # The allowlist used to determine if the strategy applies to the URL.
+        HOST_ALLOWLIST = ["cpan.metacpan.org"].freeze
+
+        # Whether the strategy can be applied to the provided URL.
+        #
+        # @param url [String] the URL to match against
+        # @return [Boolean]
+        def self.match?(url)
+          uri = URI.parse url
+          HOST_ALLOWLIST.include? uri.host
+        rescue URI::InvalidURIError
+          false
+        end
+
+        # Generates a URL and regex (if one isn't provided) and passes them
+        # to {PageMatch.find_versions} to identify versions in the content.
+        #
+        # @param url [String] the URL of the content to check
+        # @param regex [Regexp] a regex used for matching versions in content
+        # @return [Hash]
+        def self.find_versions(url, regex = nil)
+          %r{
+            /authors/id/(?<author_path>(?:[^/](?:/[^/]+){2})) # The author path (e.g. M/MI/MIYAGAWA)
+            /(?<package_path>.+) # The package path (e.g. Carton-v1.0.34.tar.gz)
+          }ix =~ url
+
+          # We need a Pathname because we've monkeypatched extname to support
+          # double extensions (e.g. tar.gz).
+          pathname = Pathname.new(package_path)
+
+          # Exteract package name
+          /^(?<package_name>.+)-v?\d+/ =~ pathname.basename(pathname.extname)
+          # Use `\.t` instead of specific tarball extensions (e.g. .tar.gz)
+          suffix = pathname.extname.sub!(/\.t(?:ar\..+|[a-z0-9]+)$/i, "\.t")
+
+          # A page containing a directory listing of the latest source tarball
+          package_dir = Pathname.new(author_path).join(pathname.dirname)
+          page_url = "https://cpan.metacpan.org/authors/id/#{package_dir}/"
+
+          # Example regex: `%r{href=.*?Carton-v?(\d+(?:\.\d+)*).t}i`
+          regex ||= /href=.*?#{package_name}-v?(\d+(?:\.\d+)*)#{Regexp.escape(suffix)}/i
+
+          Homebrew::Livecheck::Strategy::PageMatch.find_versions(page_url, regex)
+        end
```

Comment:
I think this approach is more complicated than it needs to be.

We can use `#match?` to ensure that the URL has an appropriate format and then identify the page to check by simply omitting the filename from the path. The archive file for `btparse` is found in a subdirectory in the author directory and this accounts for that situation as well.

Similarly, we can extract the parts of the path using the same regex without needing to use `Pathname`, etc. We do something similar in other strategies, so this is in line with existing approaches.

```suggestion
        def self.find_versions(url, regex = nil)
          %r{
            (?<path>/authors/id(?:/[^/]+){3,}/) # Path before the filename
            (?<prefix>[^/]+) # Filename text before the version
            -v?\d+(?:\.\d+)* # The numeric version
            (?<suffix>[^/]+) # Filename text after the version
          }ix =~ url

          # Use `\.t` instead of specific tarball extensions (e.g. .tar.gz)
          suffix.sub!(/\.t(?:ar\..+|[a-z0-9]+)$/i, "\.t")

          # The directory listing page where the archive files are found
          page_url = "https://cpan.metacpan.org#{path}"

          # Example regex: `/href=.*?brew[._-]v?(\d+(?:\.\d+)*)\.t/i`
          regex ||= /href=.*?#{prefix}[._-]v?(\d+(?:\.\d+)*)#{Regexp.escape(suffix)}/i

          Homebrew::Livecheck::Strategy::PageMatch.find_versions(page_url, regex)
        end
      end
```

Reply:
OFC you can squeeze everything into one giant pattern but there goes a question of readability of given solution 🤔 
Personally, I'm OK with your suggestion but at the same time I know a lot of people that quickly get uncomfortable when regex goes beyond one group/line 😁 

Leaving subjective topics aside I do have one minor pet issue with this:
```rb
%r{
  (?<path>(?<=/)authors/id(?:/[^/]+){3,}/) # Path before the filename
  (?<prefix>[^/]+) # Filename text before the version
  -v?\d+(?:\.\d+)* # The numeric version
  (?<suffix>[^/]+) # Filename text after the version
}ix =~ url

page_url = "https://cpan.metacpan.org/#{path}"
```
Gluing together hostname and path without slash makes me uncomfortable 😁 

Other than that it ticks all boxes: quick, elegant, and correct!

Reply:
> Gluing together hostname and path without slash makes me uncomfortable 😁

I can understand the discomfort (I feel the same way) but `URI#parse` produces a path like `/authors/id/H/HO/HOMEBREW/` (with the leading forward slash) and that feels correct to me. Manually omitting the leading forward slash simply for the sake of prettier string interpolation feels like the wrong priority to me.

If there's more support for omitting the leading forward slash, it would be better to simply move the forward slash before `authors` outside of the `path` capture group, like `/(?<path>authors/id(?:/[^/]+){3,}/)`. This would accomplish the same thing without introducing a lookbehind (which makes the regex more expensive).

Reply:
> `URI#parse` produces a path like `/authors/id/H/HO/HOMEBREW/` (with the leading forward slash) and that feels correct to me.

I'm aware of it, I'm too lazy to read RFC but it is probably by the book too.

> Manually omitting the leading forward slash simply for the sake of prettier string interpolation feels like the wrong priority to me.

😇 

> This would accomplish the same thing without introducing a lookbehind (which makes the regex more expensive).

Again I'm aware of it but... OK time to amend my commit and stop treating it like a literary form. Thanks for echoing my conscience 🙃